### ISSUE TITLE
Workbench: enable/disable vs. show/hide for inputs

### ISSUE THREAD

**Issue by dcdenu4** [natcap/invest#1601]
The toggles in the Workbench UI instantly reveal new input elements and likewise instantly hide them. This can seem to "reshuffle" inputs and since it happens instantly, it can be hard to keep track of what disappeared and what appeared, especially when inputs appear to "shift".

Example:
Open Wind Energy, select "run valuation". Toggle "use price table" on and off.

Possible solution(s):
- Maybe for some elements an accordion animation style is really what's needed. This would make sense for things like "run valuation", where a group of other elements are shown / hidden.

**Comment by davemfish** [comment]
>The toggles in the Workbench UI instantly reveal new input elements and likewise instantly hide them. This can seem to "reshuffle" inputs and since it happens instantly, it can be hard to keep track of what disappeared and what appeared, especially when inputs appear to "shift".

The toggles only enable/disable the elements, right? When they are disabled they are still on-screen, just grayed out. Are you using a feature branch with new features @dcdenu4 ?

**Comment by dcdenu4** [comment]
Hey @davemfish, you're totally right. We added a show/hide feature in the Schistosomiasis fork of InVEST. Happy to help unpack what we did there and would love to get that work upstream!

**Comment by dcdenu4** [comment]
This is relevant again because we're looking to move the schisto plugin back into compatibility with core InVEST. Hiding inputs as opposed to graying them out is crucial in that model because there are a lot of different input options which change depending on another input selection. Having inputs hidden rather than grayed out really helps clean up the UI.

I believe this is the change we made in the [schisto version of InVEST](https://github.com/natcap/schistosomiasis-invest/commit/a0f4c012cde1fe7011c8f488c3c6c483ab0a75f2#diff-dda2f63d3b4613825b7337ccfa287360274980fc017c12e52cb4835af845ac74R293). 

In that case we're just hiding instead of graying out. It likely makes sense to implement something on the InVEST side which offers more control than replacing existing functionality.

**Comment by emlys** [comment]
To keep things simple on the invest side, what if there was a "Show/hide disabled inputs" toggle in the workbench interface? That would give the user control over if/when they want to see disabled inputs, and we could use the existing enabled/disabled logic on the invest side.

It might be confusing for users if some inputs in some models disappear when they're disabled, while others remain visible and are grayed out. I don't think there's a clear distinction between the purpose of those cases that would be clear to the end user.

**Comment by dcdenu4** [comment]
>To keep things simple on the invest side, what if there was a "Show/hide disabled inputs" toggle in the workbench interface?

That's an interesting idea and likely requires the least amount of code change.

I think I still like this being a new or clarified attribute on the invest spec side because it allows the author of a model (if we think of all models as plugins) to have control over how their model is best suited for the UI. Having a toggle setting for showing / hiding disabled inputs might also be a point of confusion itself for users. Under this approach, current models would see no changes, it would simply be a new feature other models could use if it fit their needs.

**Comment by davemfish** [comment]
>I think I still like this being a new or clarified attribute on the invest spec side because it allows the author of a model (if we think of all models as plugins) to have control over how their model is best suited for the UI. Having a toggle setting for showing / hiding disabled inputs might also be a point of confusion itself for users. Under this approach, current models would see no changes, it would simply be a new feature other models could use if it fit their needs.

I agree. It seems like certain models would function best with show/hide while others would best with enable/disable. So ideally we get it right for each model ourselves rather than ask the user to choose what they like each time.

I also wonder if it makes sense to integrate this feature directly into this other one about nested inputs. https://github.com/natcap/invest/issues/1602#issuecomment-5496190083

If an input is nested within others, then maybe the Workbench uses show/hide. If an input is at the top level, it enables/disables? This is similar to Emily's suggestion, I think, except that instead of giving users control over the behavior, we decide what is best based on whether the inputs are nested or top-level.

**Comment by dcdenu4** [comment]
>If an input is nested within others, then maybe the Workbench uses show/hide. If an input is at the top level, it enables/disables? This is similar to Emily's suggestion, I think, except that instead of giving users control over the behavior, we decide what is best based on whether the inputs are nested or top-level.

I see the grouping / labeling of inputs as an orthogonal feature and am not sure it'd make sense for it to dictate show/hide | enable/disable functionality. I could imagine a use case where grouping a set of inputs and with the desire of wanting some of those inputs to act as enable/disable.

**Comment by emlys** [comment]
IMO, https://github.com/natcap/invest/issues/2726 is more relevant to this issue than #1602. Hiding all the inputs within a table could make sense if the table is disabled. Does that cover the primary use case in the schistosomiasis model?

**Comment by emlys** [comment]
I am pretty concerned that hiding some inputs according to some logic on the model side will be problematic. I think the original issue @dcdenu4 posted illustrates this perfectly:

>The toggles in the Workbench UI instantly reveal new input elements and likewise instantly hide them. This can seem to "reshuffle" inputs and since it happens instantly, it can be hard to keep track of what disappeared and what appeared, especially when inputs appear to "shift".

This was an actual issue observed after the requested feature was implemented in the schisto branch. I think it will be too confusing for users if certain inputs disappear according to some invisible conditional logic. 

Doug had suggested:
>Maybe for some elements an accordion animation style is really what's needed. This would make sense for things like "run valuation", where a group of other elements are shown / hidden.

I like this idea a lot, it gives the user control over what they want to see, and it's more intuitive than a "Show/hide disabled inputs" toggle. If we are simply trying to solve the problem of a lot of extra inputs cluttering the form, I think that can be solved independently of any custom show/hide logic.

**Comment by dcdenu4** [comment]
Thanks @emlys for your careful thought on this. I don't disagree with any of the points you brought up!

**Comment by emlys** [comment]
Just continuing to think through this... for the schisto model, in the custom invest fork, it looks like we're hiding all disabled inputs (effectively replacing the enable/disable functionality with show/hide). Is that the desired behavior, or would it be preferable to have different behavior for different inputs if that option was available?

I think it would help to get more details on what the required behavior is for each input of the schisto model, and any other use cases for this feature (@davemfish had mentioned Urban Mental Health?)

**Comment by davemfish** [comment]
@emlys I'm not sure that the UMH model has a need for this. It does have one or two conditionally-disabled inputs, but I think an early design doc may have referred to a more complicated UI than we ended up with.

 > Just continuing to think through this... for the schisto model, in the custom invest fork, it looks like we're hiding all disabled inputs (effectively replacing the enable/disable functionality with show/hide). Is that the desired behavior, or would it be preferable to have different behavior for different inputs if that option was available?

No, I don't think that is desirable. There are benefits to still being able to see a disabled input. https://github.com/natcap/invest/issues/1601#issuecomment-3299545478

My understanding is that the show/hide behavior is useful in cases where there are a large number of inputs that are sometimes not being used, so hiding them makes it easier to focus on the other inputs. That is why I suggested that show/hide might be the right design when we have a group of nested inputs like these "Population Parameters" in the Schisto case. Or if there are a bunch of other parameters required under the "Calculate Water Depth/Temperature" options.

<img width="805" height="610" alt="Image" src="https://github.com/user-attachments/assets/1986d1c9-7a14-4e59-8c05-aeed50fe6f2b" />

**Comment by emlys** [comment]
> There are benefits to still being able to see a disabled input.

I agree. I guess I'm thinking specifically for the schisto use case (which currently seems to be the only use case), is there any condition where the population parameters should be disabled but not hidden?

**Comment by dcdenu4** [comment]
>is there any condition where the population parameters should be disabled but not hidden?

No, I don't think so for this use case. In the screenshot above, if you were to select a different dropdown selection instead of "Linear", a new set of inputs would be "enabled" for "Population Parameters" and the current ones would be "disabled".

I think that "show/hide" should be separate from "enable/disable", if we were to add a "show/hide" feature. If "show/hide" were a logic driven ModelSpec param like "enable/disable", then the model developer could choose to those two features be consistent or act differently. For Schisto it would be useful for "show/hide" and "enable/disable" to share the same conditional to avoid unused inputs being passed into Args.
